org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/framework/FrameworkExtension.java

```diff
@@ -0,0 +1,277 @@
+						service = objects[0];
+					} else {
+						service = Arrays.toString(objects);
+					}
+				}
+				log.accept(service + " registered by " + reference.getBundle()
+					.getSymbolicName() + " | " + reference.getProperties());
+			});
+	}
+
+	/**
+	 * Prints the current Framework bundles and their state to the given log
+	 * consumer
+	 *
+	 * @param log for each line, this consumer will receive a string
+	 */
+	public void printComponents(Consumer<String> log) {
+		BundleContext bc = connect.framework.getBundleContext();
+		ServiceReference<ServiceComponentRuntime> reference = bc.getServiceReference(ServiceComponentRuntime.class);
+		ServiceComponentRuntime componentRuntime;
+		if (reference != null) {
+			componentRuntime = bc.getService(reference);
+		} else {
+			componentRuntime = null;
+		}
+		if (componentRuntime != null) {
+			log.accept("============ Framework Components ==================");
+			Collection<ComponentDescriptionDTO> descriptionDTOs = componentRuntime.getComponentDescriptionDTOs();
+			Comparator<ComponentConfigurationDTO> byComponentName = Comparator.comparing(dto -> dto.description.name,
+				String.CASE_INSENSITIVE_ORDER);
+			Comparator<ComponentConfigurationDTO> byComponentState = Comparator.comparingInt(dto -> dto.state);
+			descriptionDTOs.stream()
+				.flatMap(dto -> componentRuntime.getComponentConfigurationDTOs(dto)
+					.stream())
+				.sorted(byComponentState.thenComparing(byComponentName))
+				.forEachOrdered(dto -> {
+					if (dto.state == ComponentConfigurationDTO.FAILED_ACTIVATION) {
+						log.accept(toComponentState(dto.state) + " | " + dto.description.name + " | " + dto.failure);
+					} else {
+						log.accept(toComponentState(dto.state) + " | " + dto.description.name);
+					}
+					for (int i = 0; i < dto.unsatisfiedReferences.length; i++) {
+						UnsatisfiedReferenceDTO ref = dto.unsatisfiedReferences[i];
+						log.accept("\t" + ref.name + " is missing");
+					}
+				});
+			bc.ungetService(reference);
+		} else {
+			log.accept("No service component runtime installed (or started) in this framework!");
+		}
+	}
+
+	private static String toComponentState(int state) {
+		switch (state) {
+			case ComponentConfigurationDTO.ACTIVE :
+				return "ACTIVE ";
+			case ComponentConfigurationDTO.FAILED_ACTIVATION :
+				return "FAILED ";
+			case ComponentConfigurationDTO.SATISFIED :
+				return "SATISFIED ";
+			case ComponentConfigurationDTO.UNSATISFIED_CONFIGURATION :
+			case ComponentConfigurationDTO.UNSATISFIED_REFERENCE :
+				return "UNSATISFIED";
+			default :
+				return String.valueOf(state);
+		}
+	}
+
+	private static String toBundleState(int state) {
+		switch (state) {
+			case Bundle.ACTIVE :
+				return "ACTIVE   ";
+			case Bundle.INSTALLED :
+				return "INSTALLED";
+			case Bundle.RESOLVED :
+				return "RESOLVED ";
+			case Bundle.STARTING :
+				return "STARTING ";
+			case Bundle.STOPPING :
+				return "STOPPING ";
+			default :
+				return String.valueOf(state);
+		}
+	}
```

Comment:
Have a look at https://github.com/osgi/osgi-test/tree/main/org.osgi.test.common/src/main/java/org/osgi/test/common/bitmaps for a possible alternative that might avoid repeating yourself on this stuff. There is a bitmap defined for BundleState, perhaps a ComponentState one would be good too.

Reply:
Done, thanks for the hint, I think it would actually be good to have this directly in the OSGi framework :-)